### DISCUSSION TITLE
Tailwind don't recognize classes created programmatically with custom PostCSS plugin

### DISCUSSION THREAD

**Discussion by alexanderkalachev** [tailwindlabs/tailwindcss#12107] [Help] [answered]
I am writing PostCSS plugin to create custom css classes and using it alongside Tailwind.
Everything works perfect except when I am trying to define a class in components layer in css file with global styles.
First of all, is not processing @apply rules inside those classes, and also if I'm trying to call for created class in other @apply rule, it return an error:
"The `typo-caps-3xl` class does not exist. If `typo-caps-3xl` is a custom class, make sure it is defined within a `@layer` directive."

postcss.config.js
```
module.exports = {
    plugins: [
        require.resolve("postcss-import"),
        require.resolve("my-custom-plugin"),
        require.resolve("tailwindcss"),
        require.resolve("autoprefixer")
    ]
};
```

globals.css
```
@import "some-css.css";
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer base {
    ...

    h1 {
        @apply typo-caps-3xl;
    }
}

@layer components {
    @typo-components; // Anchor where custom classes are placed

    ...
    }
}
```

generated css
```
.typo-caps-3xl {
    @apply calc-fontsize;
    
    ...
}
```

**Comment by wongjn** [comment]
Consider checking the rules that `@typo-components` expands to are still placed inside a relevant `@layer` directive after processing by your PostCSS plugin. Otherwise, consider providing a remotely-hosted git repository that reproduces the unexpected behavior.

**Reply by alexanderkalachev** [reply]
I'm simply replacing the node @typo-components with a a root with rules. But it seems like you are right, even that those rules are in the required place in CSS, Tailwind might see it outside the components layer.
But I don't know why. I have other classes inside the same layer and Tailwind are capable to see that they are still in the same layer even after my plugin processes the file.
hmmm... I'm new to PostCSS and it might be a question related to it, rather than Tailwind.

**Reply by alexanderkalachev** [reply]
Ok, now I'm certain that rules are placed inside the right layer.

**Reply by wongjn** [reply]
Consider providing a remotely-hosted git repository that reproduces the unexpected behavior.

**Reply by alexanderkalachev** [reply]
Ok, I created a repository with similar setup that I have.
https://github.com/alexanderkalachev/postcss-generate-classes-in-tw-layers
It's Next.js app with PostCSS and Tailwind.
You can find two css files inside 'app' directory: globals.css and partial.css
I use postcss-import to include partial.css in globals.css.
Then I generate two additional classes with custom plugin in 'plugins/custom-plugin.js'.

Then inside globals.css I trying to access different classes inside components layer. Each class represent a way to add a component that doesn't work.

Thank you! Please, let me know if you can spot my mistake.

**Reply by wongjn** [reply, marked as answer]
So it seems like this line runs before your custom PostCSS plugin:

https://github.com/tailwindlabs/tailwindcss/blob/0bd81a06c499be58bf87ca16481333823d86e828/src/plugin.js#L35

Which eventually calls:

https://github.com/tailwindlabs/tailwindcss/blob/0bd81a06c499be58bf87ca16481333823d86e828/src/lib/setupContextUtils.js#L713-L719

But your own plugin hasn't run yet so your custom at-rules haven't been expanded yet.

To work-around this, you could consider refactoring your PostCSS plugin rules into a Tailwind plugin instead.

**Reply by alexanderkalachev** [reply]
Will it be possible to create and assign css variable with Tailwind plugin? This is the main reason I didn't choose this way in a first place.

**Reply by wongjn** [reply]
Yes, it would be possible.

**Reply by alexanderkalachev** [reply]
Thank you! I will look into it.
But still it's a shame that this particular task cannot be solved.

**Reply by alexanderkalachev** [reply]
Ok, that's actually even easier! Thank you for inspiring me to look into that!
